### PR TITLE
chore: refactor block and state api utils

### PR DESCRIPTION
**Pull request by wemeetagain**
**Motivation**

- simplify the diff in https://github.com/ChainSafe/lodestar/pull/6033

**Description**

- refactor block and state api utils (`resolveBlockId` and `resolveStateId`)
  - now the `resolve*` functions return a concrete slot or root which can be fetched using the chain
- fetching blocks and states in the API is now done with `getBlockResponse` and `getStateResponse`
- in preparation for #6033, a separate `getStateResponseWithRegen` is created with unimplemented logic and a special return type.
  - This is a separate function to simplify consumer code (not have to handle the Uint8Array case) and to prevent accidental usage of a dangerous / expensive operation.

### CODE REVIEW

File: packages/beacon-node/src/api/impl/beacon/blocks/utils.ts [hunk 1]
```diff
@@ -22,50 +23,53 @@ export function toBeaconHeaderResponse(
...
-  }
-
-  return res;
-}
-
-async function resolveBlockIdOrNull(
-  chain: IBeaconChain,
-  blockId: routes.beacon.BlockId
-): Promise<{block: SignedBeaconBlock; executionOptimistic: boolean; finalized: boolean} | null> {
+export function resolveBlockId(forkChoice: IForkChoice, blockId: routes.beacon.BlockId): string | number {
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/blocks/utils.ts]: could reuse type from api here

```suggestion
export function resolveBlockId(forkChoice: IForkChoice, blockId: routes.beacon.BlockId): routes.beacon.BlockId {
```

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 2]
```diff
@@ -3,51 +3,32 @@ import {FAR_FUTURE_EPOCH, GENESIS_SLOT} from "@lodestar/params";
...
-
-  return stateRes;
-}
-
-async function resolveStateIdOrNull(
-  chain: IBeaconChain,
-  stateId: routes.beacon.StateId,
-  opts?: StateGetOpts
-): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean} | null> {
+export function resolveStateId(forkChoice: IForkChoice, stateId: routes.beacon.StateId): string | number {
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: same as for blocks
```suggestion
export function resolveStateId(forkChoice: IForkChoice, stateId: routes.beacon.StateId): routes.beacon.StateId {
```

File: packages/beacon-node/src/api/impl/beacon/blocks/utils.ts [hunk 3]
```diff
@@ -22,50 +23,53 @@ export function toBeaconHeaderResponse(
...
+  if (typeof rootOrSlot === "string") {
+    res = await chain.getBlockByRoot(rootOrSlot);
+  } else {
+    res = await chain.getCanonicalBlockAtSlot(rootOrSlot);
+  }
+  if (!res) {
+    throw new ApiError(404, `No block found for id '${blockId}'`);
+  }
+
+  return res;
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/blocks/utils.ts]: what about

```suggestion
  const rootOrSlot = resolveBlockId(chain.forkChoice, blockId);

  const res =
    typeof rootOrSlot === "string"
      ? await chain.getBlockByRoot(rootOrSlot)
      : await chain.getCanonicalBlockAtSlot(rootOrSlot);

  if (!res) {
    throw new ApiError(404, `No block found for id '${blockId}'`);
  }

  return res;
```

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 4]
```diff
@@ -3,51 +3,32 @@ import {FAR_FUTURE_EPOCH, GENESIS_SLOT} from "@lodestar/params";
...
+    return head.stateRoot;
   }
 
   if (stateId === "genesis") {
-    return chain.getStateBySlot(GENESIS_SLOT, opts);
+    return GENESIS_SLOT;
   }
 
   if (stateId === "finalized") {
-    const checkpoint = chain.forkChoice.getFinalizedCheckpoint();
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts:40]: if we just return the state root, isn't it cheaper to do, or is `rootHex` not the state root?
```ts
const checkpoint = forkChoice.getFinalizedCheckpoint();
return checkpoint.rootHex;
```

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 5]
```diff
@@ -3,51 +3,32 @@ import {FAR_FUTURE_EPOCH, GENESIS_SLOT} from "@lodestar/params";
...
   if (stateId === "justified") {
-    const checkpoint = chain.forkChoice.getJustifiedCheckpoint();
-    return chain.getStateByCheckpoint(checkpoint);
+    const block = forkChoice.getJustifiedBlock();
+    return block.stateRoot;
   }
 
   if (typeof stateId === "string" && stateId.startsWith("0x")) {
-    return chain.getStateByStateRoot(stateId, opts);
+    return stateId as string;
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: Not needed

```suggestion
    return stateId;
```

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 6]
```diff
@@ -56,7 +37,50 @@ async function resolveStateIdOrNull(
...
+export async function getStateResponse(
+  chain: IBeaconChain,
+  stateId: routes.beacon.StateId
+): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean}> {
+  const rootOrSlot = resolveStateId(chain.forkChoice, stateId);
+
+  let state = null;
+  if (typeof rootOrSlot === "string") {
+    state = await chain.getStateByStateRoot(rootOrSlot);
+  } else if (typeof rootOrSlot === "number") {
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: why is this extra `typeof` check here?

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 7]
```diff
@@ -56,7 +37,50 @@ async function resolveStateIdOrNull(
...
+  return state;
+}
+
+export async function getStateResponseWithRegen(
+  chain: IBeaconChain,
+  stateId: routes.beacon.StateId
+): Promise<{state: BeaconStateAllForks | Uint8Array; executionOptimistic: boolean; finalized: boolean}> {
+  const rootOrSlot = resolveStateId(chain.forkChoice, stateId);
+
+  let state: {state: BeaconStateAllForks | Uint8Array; executionOptimistic: boolean; finalized: boolean} | null = null;
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: Maybe we can create a type like this

```ts
type StateWithMetadata = {state: BeaconStateAllForks | Uint8Array; executionOptimistic: boolean; finalized: boolean};
```

and same for block

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 8]
```diff
@@ -56,7 +37,50 @@ async function resolveStateIdOrNull(
...
+  return blockSlot;
+}
+
+export async function getStateResponse(
+  chain: IBeaconChain,
+  stateId: routes.beacon.StateId
+): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean}> {
+  const rootOrSlot = resolveStateId(chain.forkChoice, stateId);
+
+  let state = null;
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts]: here it's called state but might wanna call it `res` as above for blocks

File: packages/beacon-node/src/util/sszBytes.ts [hunk 9]
```diff
@@ -6,6 +6,17 @@ import {BYTES_PER_FIELD_ELEMENT, FIELD_ELEMENTS_PER_BLOB} from "@lodestar/params
...
 export type AttDataBase64 = string;
 
+// class BeaconState(Container):
+//   genesis_time: uint64              - 8
+//   genesis_validators_root: bytes32  - 32
+//   slot: uint64
+//   ...
+const STATE_SLOT_OFFSET = 40;
+
+export function getSlotFromBeaconStateSerialized(data: Uint8Array): Slot {
```
Comment (nflaig) [inline packages/beacon-node/src/util/sszBytes.ts]: can we use existing `getStateSlotFromBytes`?

https://github.com/ChainSafe/lodestar/blob/c6369ee5fe2c41f922e36afaedd7db4d8a69b15f/packages/beacon-node/src/util/multifork.ts#L44

File: packages/beacon-node/src/api/impl/beacon/blocks/utils.ts (hunk 1 above)
Comment (wemeetagain) [reply packages/beacon-node/src/api/impl/beacon/blocks/utils.ts]: hah, I'd prefer not to do that. Structurally the same, semantically not the same

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts (hunk 4 above)
Comment (wemeetagain) [reply packages/beacon-node/src/api/impl/beacon/state/utils.ts:40]: rootHex is the block root

File: packages/beacon-node/src/util/sszBytes.ts (hunk 9 above)
Comment (wemeetagain) [reply packages/beacon-node/src/util/sszBytes.ts]: ah good call

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts (hunk 6 above)
Comment (wemeetagain) [reply packages/beacon-node/src/api/impl/beacon/state/utils.ts]: sloppiness, will update as in block

File: packages/beacon-node/src/api/impl/beacon/blocks/utils.ts (hunk 1 above)
Comment (nflaig) [reply packages/beacon-node/src/api/impl/beacon/blocks/utils.ts]: ah nvm this comment, block id also has `"head" | "genesis" | "finalized" | "justified"`, so yeah, it's not the same

**Review by nflaig — APPROVED** [review]
LGTM

**Comment by nflaig** [comment]
I think node 22.5.0 blows up the CI build, might need to pin it to 22.4

File: packages/beacon-node/src/api/impl/beacon/state/utils.ts [hunk 10]
```diff
@@ -1,53 +1,31 @@
...
-  chain: IBeaconChain,
-  stateId: routes.beacon.StateId,
-  opts?: StateGetOpts
-): Promise<{state: BeaconStateAllForks; executionOptimistic: boolean; finalized: boolean} | null> {
+export function resolveStateId(forkChoice: IForkChoice, stateId: routes.beacon.StateId): RootHex | Slot {
   if (stateId === "head") {
-    // TODO: This is not OK, head and headState must be fetched atomically
-    const head = chain.forkChoice.getHead();
-    const headState = chain.getHeadState();
-    return {state: headState, executionOptimistic: isOptimisticBlock(head), finalized: false};
```
Comment (nflaig) [inline packages/beacon-node/src/api/impl/beacon/state/utils.ts:32]: @wemeetagain wondering here, we got the head state from `getHeadState()` which called `getClosestHeadState`

so we would previouly do this

https://github.com/ChainSafe/lodestar/blob/fed08fe5107a809d8862494ae6a86ebc6f2f0ee9/packages/beacon-node/src/chain/regen/queued.ts#L142

which queried both `blockStateCache` and `checkpointStateCache`



while now we would just query `blockStateCache`

https://github.com/ChainSafe/lodestar/blob/fed08fe5107a809d8862494ae6a86ebc6f2f0ee9/packages/beacon-node/src/chain/regen/queued.ts#L79-L80

Comment (twoeths) [reply packages/beacon-node/src/api/impl/beacon/state/utils.ts:32]: @nflaig , the change actually made it better. Previously we also queried the head state pulled up to next epoch in the `checkpointStateCache` which is only necessary for block processing while the "getHeadState()" api only wants the real head state. So it's better to only query `blockStateCache` in this case

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.21.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.21.0) :tada:

**Pull request merged**